Tests/MathLib/mechanics_invariants.cpp

```diff
@@ -0,0 +1,286 @@
+                  KelvinToTensorSize<4>::value>
+        m;
+    // The z-component is ignored, otherwise the matrix must be 3x3.
+    m << v[0], v[3] / std::sqrt(2.), v[3] / std::sqrt(2.), v[1];
+    return m;
+}
+
+template <>
+Eigen::Matrix<double, KelvinToTensorSize<6>::value,
+              KelvinToTensorSize<6>::value>
+    kelvinToTensor(KelvinVector<6> const& v)
+{
+    Eigen::Matrix<double, KelvinToTensorSize<6>::value,
+                  KelvinToTensorSize<6>::value>
+        m;
+    m << v[0], v[3] / std::sqrt(2.), v[5] / std::sqrt(2.), v[3] / std::sqrt(2.),
+        v[1], v[4] / std::sqrt(2.), v[5] / std::sqrt(2.), v[4] / std::sqrt(2.),
+        v[2];
+    return m;
+}
+
+template <int Size>
+KelvinVector<Size> tensorToKelvin(
+    Eigen::Matrix<double, KelvinToTensorSize<Size>::value,
+                  KelvinToTensorSize<Size>::value> const& m);
+
+template <>
+KelvinVector<1> tensorToKelvin(
+    Eigen::Matrix<double, KelvinToTensorSize<1>::value,
+                  KelvinToTensorSize<1>::value> const& m)
+{
+    Eigen::Matrix<double, 1, 1> v;
+    v << m[0];
+    return v;
+}
+
+template <>
+KelvinVector<4> tensorToKelvin(
+    Eigen::Matrix<double, KelvinToTensorSize<4>::value,
+                  KelvinToTensorSize<4>::value> const& m)
+{
+    KelvinVector<4> v;
+    // 0 is for the z-component, otherwise the matrix must be 3x3.
+    v << m(0, 0), m(1, 1), 0, m(0, 1) * std::sqrt(2.);
+    return v;
+}
+
+template <>
+KelvinVector<6> tensorToKelvin(
+    Eigen::Matrix<double, KelvinToTensorSize<6>::value,
+                  KelvinToTensorSize<6>::value> const& m)
+{
+    KelvinVector<6> v;
+    v << m(0, 0), m(1, 1), m(2, 2), m(0, 1) * std::sqrt(2.),
+        m(1, 2) * std::sqrt(2.), m(0, 2) * std::sqrt(2.);
+    return v;
+}
+
+struct MathLibMechanicsInvariants1 : public ::testing::Test
+{
+    static const int size = 1;
+    ac::randomEigenMatrixGenerator<double, size, 1> kelvinVectorGenerator;
+
+    ac::gtest_reporter gtest_reporter;
+};
+
+struct MathLibMechanicsInvariants4 : public ::testing::Test
+{
+    static const int size = 4;
+    ac::randomEigenMatrixGenerator<double, size, 1> kelvinVectorGenerator;
+
+    ac::gtest_reporter gtest_reporter;
+};
+
+struct MathLibMechanicsInvariants6 : public ::testing::Test
+{
+    static const int size = 6;
+    ac::randomEigenMatrixGenerator<double, size, 1> kelvinVectorGenerator;
+
+    ac::gtest_reporter gtest_reporter;
+};
+
+KelvinVector<4> setZcomponentZero(KelvinVector<4> const& v, std::size_t)
+{
+    auto result = v;
+    result[2] = 0;
+    return result;
+};
+
+//
+// Self-test of Kelvin to tensor and back conversion, which should be identity
+// up to numerical errors.
+//
+
+TEST_F(MathLibMechanicsInvariants1, SelfTestMappingKelvinToTensor)
+{
+    auto f = [](KelvinVector<1> const& v) {
+        return v == tensorToKelvin<1>(kelvinToTensor(v));
+    };
+
+    ac::check<KelvinVector<1>>(
+        f, 1000, ac::make_arbitrary(kelvinVectorGenerator), gtest_reporter);
```

Comment:
In debug 30ms for selftest, 30ms for det, and 350ms for inv. for the size 6.
All ' MathLibMechanicsInvariants6.*' tests run in 422ms including testrunner startup/shutdown.